Server.py

```diff
@@ -12,36 +12,43 @@
 serverSocket.listen(1) 
 list_of_connections = {}
 
+def user_login_server(username,password,conn): 
+    userObjList = structure.Users.values()
+    for userObj in userObjList: 
+        if (userObj.user_id == username) and (userObj.password == password):
+            conn.send("match")
+        else:
+            conn.send("nomatch")
 def add_user_server(user_id, new_user_id, chatroom_name, conn): 
+  
+  #hard-coded version 
   chatid = structure.Chatnames.get(chatroom_name)
+  currentUserObj = structure.Users.get(user_id)
   if chatid == None:
-    conn.send("Chatroom does not exist.") #is the 'b' a typo?
-  current_chat = structure.Chats.get(chatid) #Chat object
-  user_list = structure.demo_chat.chat_users
-  if new_user_id in structure.demo_chat.chat_users: #replace with current_chat later
-    conn.send("User is already in chat")
+    conn.send("Chatroom does not exist.".encode())  
+  this_chat_obj = structure.Chats.get(chatid)
+  
+  if new_user_id in this_chat_obj.chat_users:  
+    conn.send("User is already in chat".encode())
   else: 
-    #insert user id into current chat's user list
-    structure.demo_chat.chat_users.append(new_user_id) 
-    #can we store full names in the User object? Then I could have the user's name instead of username
+    this_chat_obj.chat_users.append(new_user_id) #how to add globally? 
     message = ("Added new user {uid} to the chat").format(uid = new_user_id)
-     #initialize chat's last_pushed_time to current time (for new user)
     structure.Chat_user3.last_pushed_time = datetime.datetime.now()  
     #add chat_id to user's list of chats
     structure.Chat_user3.chats.append(chatid)
      #send confirmation message back to the client 
-    conn.send(message) 
+    conn.send(message.encode()) 
      #send_message("wendy added to the chatroom")
-    send_message_server(user_id, chatroom_name, message, conn)
-    #when a new user is added to a chat, update chat's name? 
 
+    # send_message_server(user_id, chatroom_name, message, conn)
+    #when a new user is added to a chat, update chat's name? 
 
 def delete_user_server(user_id, user_begone_id, chatroom_name,conn):
   if user_begone_id in structure.demo_chat.chat_users: #replace with current_chat later
     chatid = structure.Chatnames.get(chatroom_name)
     # Delete the user sublist in the users dict (in the chat hash table)
     structure.demo_chat.chat_users.remove(structure.Chat_user_obj('Leah')) #probably a better way to do this
-       # Remove the chat_id from the user’s list of chats (in the user table)
+       # Remove the chat id from the user’s list of chats (in the user table)
```

Comment:
`SyntaxError: Non-ASCII character '\xe2' in file Server.py on line 51, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details` appears here and I'm not quite sure how to fix it – I've encoded messages sent back to the server, but that moved where in my code the error is occurring.